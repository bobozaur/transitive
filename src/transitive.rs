use darling::{FromAttributes, FromDeriveInput};
use proc_macro2::TokenStream;
use quote::ToTokens;
use syn::{Attribute, DeriveInput, Ident, Result as SynResult};

use crate::{
    fallible::{TransitiveTryFrom, TransitiveTryInto},
    infallible::{TransitiveFrom, TransitiveInto},
};

#[derive(FromDeriveInput)]
#[darling(forward_attrs(transitive))]
struct DeriveData {
    ident: Ident,
    attrs: Vec<Attribute>,
}

enum TransitiveAttr {
    From(TransitiveFrom),
    Into(TransitiveInto),
    TryFrom(TransitiveTryFrom),
    TryInto(TransitiveTryInto),
}

pub struct TransitiveIdent<'a, T> {
    pub ident: &'a Ident,
    pub data: T,
}

impl<'a, T> TransitiveIdent<'a, T> {
    pub fn new(ident: &'a Ident, data: T) -> Self {
        Self { ident, data }
    }
}

impl FromAttributes for TransitiveAttr {
    fn from_attributes(attrs: &[Attribute]) -> darling::Result<Self> {
        TransitiveFrom::from_attributes(attrs)
            .map(TransitiveAttr::From)
            .or_else(|_| TransitiveInto::from_attributes(attrs).map(TransitiveAttr::Into))
            .or_else(|_| TransitiveTryFrom::from_attributes(attrs).map(TransitiveAttr::TryFrom))
            .or_else(|_| TransitiveTryInto::from_attributes(attrs).map(TransitiveAttr::TryInto))
    }
}

impl ToTokens for TransitiveIdent<'_, TransitiveAttr> {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        match &self.data {
            TransitiveAttr::From(from) => TransitiveIdent::new(self.ident, from).to_tokens(tokens),
            TransitiveAttr::Into(into) => TransitiveIdent::new(self.ident, into).to_tokens(tokens),
            TransitiveAttr::TryFrom(try_from) => {
                TransitiveIdent::new(self.ident, try_from).to_tokens(tokens)
            }
            TransitiveAttr::TryInto(try_into) => {
                TransitiveIdent::new(self.ident, try_into).to_tokens(tokens)
            }
        }
    }
}

pub fn transitive_impl(input: DeriveInput) -> SynResult<TokenStream> {
    let DeriveData { ident, attrs } = DeriveData::from_derive_input(&input)?;
    let mut output = TokenStream::new();

    for attr in attrs {
        let attr = TransitiveAttr::from_attributes(&[attr])?;
        let attr = TransitiveIdent::new(&ident, attr);
        output.extend(attr.into_token_stream())
    }

    Ok(output)
}
