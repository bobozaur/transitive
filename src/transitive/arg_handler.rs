use std::str::FromStr;

use proc_macro2::TokenStream;
use quote::quote;
use syn::{Ident, Path, Result as SynResult};

use super::{ArgListType, MinimalAttrArgs};

pub trait ArgHandler {
    fn conv_func_name(&self) -> &str;

    fn create_bulk_impl(
        &self,
        name: &Ident,
        stmts: TokenStream,
        first: Path,
        last: Path,
        second_last: Option<Path>,
    ) -> TokenStream;

    fn create_pair_impl(&self, name: &Ident, first: &Path, last: &Path) -> TokenStream;

    fn stmt_end(&self) -> TokenStream {
        TokenStream::new()
    }

    fn make_impl(&self, name: &Ident, args: ArgListType) -> SynResult<TokenStream> {
        match args {
            ArgListType::Simple(args) => self.make_bulk_impl(name, args),
            ArgListType::All(args) => self.make_pair_impls(name, args),
        }
    }

    /// Processes an argument list considering regular behavior
    /// of implementing the trait only between source type and target type.
    fn make_bulk_impl(
        &self,
        name: &Ident,
        args: MinimalAttrArgs,
    ) -> SynResult<TokenStream> {
        let MinimalAttrArgs {
            first,
            mut last,
            iter,
        } = args;

        let stmt_end = self.stmt_end();
        let func = TokenStream::from_str(self.conv_func_name())?;

        let mut second_last = None;

        // Create the buffer and store the minimum amount of statements.
        let mut stmts = TokenStream::new();
        stmts.extend(quote! {let interm = #first::#func(val)#stmt_end;});
        stmts.extend(quote! {let interm = #last::#func(interm)#stmt_end;});

        // Store other statements, if any
        for param in iter {
            second_last = Some(last);
            last = param?;
            stmts.extend(quote! {let interm = #last::#func(interm)#stmt_end;});
        }

        // Generate code
        let expanded = self.create_bulk_impl(name, stmts, first, last, second_last);
        Ok(expanded)
    }

    /// Processes an argument list considering the enhanced behavior
    /// of implementing the trait between all transitions from either
    /// one source and multiple targets or multiple targets and once source,
    /// depending on the trait.
    fn make_pair_impls(
        &self,
        name: &Ident,
        args: MinimalAttrArgs,
    ) -> SynResult<TokenStream> {
        let MinimalAttrArgs {
            mut first,
            mut last,
            iter,
        } = args;

        // Create the buffer and store the first impl.
        let mut impls = TokenStream::new();
        impls.extend(self.create_pair_impl(name, &first, &last));

        // Create and store other impls, if any
        for param in iter {
            first = last;
            last = param?;
            impls.extend(self.create_pair_impl(name, &first, &last));
        }

        Ok(impls)
    }
}
