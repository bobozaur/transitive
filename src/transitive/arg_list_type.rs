use syn::{spanned::Spanned, Error, Meta, MetaList, NestedMeta, Result as SynResult};

use super::{minimal_args::MinimalAttrArgs, RawArgList, ALL};

/// Enum diferentiating between the simple or enhance
/// transitive behavior.
pub enum ArgListType {
    Simple(MinimalAttrArgs),
    All(MinimalAttrArgs),
}

/// Conversion between [`RawArgList`] and [`ArgListType`],
/// used to determine how to process the arguments provided.
/// The attribute must always have arguments.
///
/// If there's only one argument, we assume that is the `all()`
/// argument which contains a list in itself, annotating to use
/// the enhanced behavior (multiple sources or targets).
///
/// Otherwise, consider the arguments as a list of type to transition
/// through from only one source to only one target.
impl TryFrom<RawArgList> for ArgListType {
    type Error = Error;

    fn try_from(value: RawArgList) -> SynResult<Self> {
        match value.len() {
            0 => Err(Error::new(value.span(), "missing arguments")),
            1 => Ok(value.into_iter().next().unwrap().try_into()?),
            _ => Ok(Self::Simple(value.try_into()?)),
        }
    }
}

impl TryFrom<NestedMeta> for ArgListType {
    type Error = Error;

    fn try_from(value: NestedMeta) -> Result<Self, Self::Error> {
        let NestedMeta::Meta(Meta::List(list)) = value else {
            return Err(Error::new(value.span(), "must provide a list of types"))
        };

        list.try_into()
    }
}

impl TryFrom<MetaList> for ArgListType {
    type Error = Error;

    fn try_from(value: MetaList) -> Result<Self, Self::Error> {
        if value.path.is_ident(ALL) {
            Ok(Self::All(value.nested.try_into()?))
        } else {
            Err(Error::new(value.path.span(), "unknown argument"))
        }
    }
}
