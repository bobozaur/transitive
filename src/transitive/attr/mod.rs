mod idented;

use darling::FromAttributes;
use syn::Attribute;

use super::{
    fallible::{TransitiveTryFrom, TransitiveTryInto},
    infallible::{TransitiveFrom, TransitiveInto},
};

pub use idented::AttrWithIdent;

pub enum TransitiveAttr {
    From(TransitiveFrom),
    Into(TransitiveInto),
    TryFrom(TransitiveTryFrom),
    TryInto(TransitiveTryInto),
}

impl FromAttributes for TransitiveAttr {
    fn from_attributes(attrs: &[Attribute]) -> darling::Result<Self> {
        TransitiveFrom::from_attributes(attrs)
            .map(TransitiveAttr::From)
            .or_else(|_| TransitiveInto::from_attributes(attrs).map(TransitiveAttr::Into))
            .or_else(|_| TransitiveTryFrom::from_attributes(attrs).map(TransitiveAttr::TryFrom))
            .or_else(|_| TransitiveTryInto::from_attributes(attrs).map(TransitiveAttr::TryInto))
    }
}
