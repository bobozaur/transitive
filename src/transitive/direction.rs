use std::marker::PhantomData;

use quote::__private::ext::RepToTokensExt;
use syn::{
    parse::{Parse, ParseStream},
    spanned::Spanned,
    Error, Ident, Meta, MetaList, NestedMeta, Result as SynResult,
};

use super::{arg_list_type::ArgListType, direction_handler::DirectionKind, VALID_ARGS};

pub enum Direction {
    From(ArgListType),
    Into(ArgListType),
}

pub struct DirectionWrapper<K>
where
    K: DirectionKind,
{
    direction: Option<Direction>,
    marker: PhantomData<fn() -> K>,
}

impl<K> DirectionWrapper<K>
where
    K: DirectionKind,
{
    pub fn into_inner(self) -> Option<Direction> {
        self.direction
    }

    fn direction_from(args: ArgListType) -> Self {
        Self {
            direction: Some(Direction::From(args)),
            marker: PhantomData,
        }
    }

    fn direction_into(args: ArgListType) -> Self {
        Self {
            direction: Some(Direction::Into(args)),
            marker: PhantomData,
        }
    }

    fn is_valid_arg(arg: &Ident) -> bool {
        VALID_ARGS.iter().find(|v| arg == v).next().is_some()
    }
}

impl<K> Default for DirectionWrapper<K>
where
    K: DirectionKind,
{
    fn default() -> Self {
        Self {
            direction: Default::default(),
            marker: Default::default(),
        }
    }
}

impl<K> TryFrom<NestedMeta> for DirectionWrapper<K>
where
    K: DirectionKind,
{
    type Error = Error;

    fn try_from(value: NestedMeta) -> Result<Self, Self::Error> {
        let NestedMeta::Meta(Meta::List(list)) = value else {
            return Err(Error::new(value.span(), "must provide a direction"))
        };

        list.try_into()
    }
}

impl<K> TryFrom<MetaList> for DirectionWrapper<K>
where
    K: DirectionKind,
{
    type Error = Error;

    fn try_from(value: MetaList) -> Result<Self, Self::Error> {
        match value.path.get_ident() {
            Some(i) if i == K::arg_from() => Ok(Self::direction_from(value.nested.try_into()?)),
            Some(i) if i == K::arg_into() => Ok(Self::direction_into(value.nested.try_into()?)),
            Some(i) if Self::is_valid_arg(i) => Ok(Self::default()),
            Some(i) => Err(Error::new(i.span(), format!("unknown argument {i}"))),
            None => Err(Error::new(value.path.span(), "missing direction argument")),
        }
    }
}

/// Implementing [`Parse`] to directly parse the given attribute.
/// The attribute must always have arguments.
///
/// The primary arguments decide the direction: `from()` or `into()`.
/// These contain a types list, optionally encapsulated in
/// `all()`
///
/// The type lists must have at least two arguments, otherwise
/// there's no transition to do.
impl<K> Parse for DirectionWrapper<K>
where
    K: DirectionKind,
{
    fn parse(input: ParseStream) -> SynResult<Self> {
        let nested = NestedMeta::parse(input)?;
        nested.try_into()
    }
}
