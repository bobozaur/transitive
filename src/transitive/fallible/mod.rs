mod try_from;
mod try_into;

use syn::{
    parse::{discouraged::Speculative, Parse, ParseStream},
    punctuated::Punctuated,
    spanned::Spanned,
    Error as SynError, Ident, Result as SynResult, Token, Type,
};
pub use try_from::TryTransitionFrom;
pub use try_into::TryTransitionInto;

/// A path list that may contain a custom error type.
pub struct FalliblePathList {
    /// First type in the transitive conversion. ie. `A` in
    /// `#[transitive(try_from(A, B, C, D, E))]`
    first_type: Type,
    /// Intermediate types for the transitive conversion. ie. `[B, .., D]` in
    /// `#[transitive(try_from(A, B, C, D, E))]`
    intermediate_types: Vec<Type>,
    /// Last type in the transitive conversion. ie. `E` in
    /// `#[transitive(try_from(A, B, C, D, E))]`
    last_type: Type,
    error: Option<Type>,
}

impl Parse for FalliblePathList {
    fn parse(input: ParseStream) -> SynResult<Self> {
        let attr_list = Punctuated::<Item, Token![,]>::parse_terminated(input)?;

        let mut attr_list_iter = attr_list.iter().cloned();
        let (first_type, mut last_type) = match (attr_list_iter.next(), attr_list_iter.next()) {
            (Some(Item::Type(first)), Some(Item::Type(last))) => (first, last),
            _ => {
                static TOO_FEW_TYPES_ERR_MSG: &str = "at least two types required";
                return Err(SynError::new(attr_list.span(), TOO_FEW_TYPES_ERR_MSG));
            }
        };

        let mut intermediate_types = Vec::with_capacity(attr_list.len());
        let mut error = None;

        for attr in attr_list_iter {
            match attr {
                Item::Type(ty) if error.is_some() => {
                    let msg = "types not allowed after 'error'";
                    return Err(SynError::new_spanned(ty, msg));
                }
                // Just a regular type path in the conversion path
                Item::Type(ty) => {
                    intermediate_types.push(last_type);
                    last_type = ty;
                }
                Item::Error(err) if error.is_some() => {
                    let msg = "'error' not allowed multiple times";
                    return Err(SynError::new_spanned(err, msg));
                }
                // Custom error, but must check that it's a type path
                Item::Error(err) => error = Some(err),
            }
        }

        let output = Self {
            first_type,
            intermediate_types,
            last_type,
            error,
        };

        Ok(output)
    }
}

/// An item in the parameters list of an attribute.
#[derive(Clone)]
pub enum Item {
    Type(Type),
    Error(Type),
}

impl Parse for Item {
    fn parse(input: ParseStream) -> SynResult<Self> {
        let fork = input.fork();
        // Parse the ident name and the equal sign after it
        let res = fork
            .parse::<Ident>()
            .and_then(|ident| fork.parse::<Token![=]>().map(|_| ident));

        match res {
            // We got an `error = MyType` argument
            Ok(path) if path == "error" => {
                input.advance_to(&fork);
                input.parse().map(Self::Error)
            }
            // Try to parse anything else as a type in the path list
            _ => input.parse().map(Self::Type),
        }
    }
}

impl quote::ToTokens for Item {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        match self {
            Self::Type(ty) | Self::Error(ty) => ty.to_tokens(tokens),
        }
    }
}
