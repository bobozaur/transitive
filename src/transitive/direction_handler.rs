use proc_macro2::TokenStream;
use syn::Result as SynResult;
use syn::{punctuated::Punctuated, Attribute, DeriveInput, Ident};

use super::direction::DirectionWrapper;
use super::{arg_handler::ArgHandler, direction::Direction, RawArgList, TRANSITIVE};

pub trait DirectionHandler {
    type IntoHandler: ArgHandler;
    type FromHandler: ArgHandler;
    type Kind: DirectionKind;

    fn handler_into(&self) -> Self::IntoHandler;

    fn handler_from(&self) -> Self::FromHandler;

    fn impl_with_direction(
        &self,
        name: &Ident,
        direction: DirectionWrapper<Self::Kind>,
    ) -> SynResult<Option<TokenStream>> {
        let Some(direction) = direction.into_inner() else {
            return Ok(None)
        };

        match direction {
            Direction::Into(args) => self.handler_into().make_impl(name, args).map(Some),
            Direction::From(args) => self.handler_from().make_impl(name, args).map(Some),
        }
    }

    /// Determines the attribute kind and processes it accordingly.
    fn parse_transitive_attr(attr: Attribute) -> Option<SynResult<RawArgList>> {
        if attr.path.is_ident(TRANSITIVE) {
            Some(attr.parse_args_with(Punctuated::parse_terminated))
        } else {
            None
        }
    }

    /// Parse input and processes attributes with the provided parameters.
    fn generate_tokens(&self, input: DeriveInput) -> SynResult<TokenStream> {
        let name = input.ident;
        let mut expanded = TokenStream::new();

        let attr_iter = input
            .attrs
            .into_iter()
            .filter_map(Self::parse_transitive_attr);

        for meta_iter in attr_iter {
            for nested_meta in meta_iter? {
                let direction = DirectionWrapper::<Self::Kind>::try_from(nested_meta)?;
                let tokens = self.impl_with_direction(&name, direction)?;

                if let Some(tokens) = tokens {
                    expanded.extend(tokens)
                };
            }
        }

        Ok(expanded)
    }
}

pub trait DirectionKind {
    fn arg_from() -> &'static str;

    fn arg_into() -> &'static str;
}
