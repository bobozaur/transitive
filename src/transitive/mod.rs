mod fallible;
mod infallible;

use fallible::{TryTransitionFrom, TryTransitionInto};
use infallible::{TransitionFrom, TransitionInto};
use proc_macro2::TokenStream;
use quote::ToTokens;
use syn::{
    parse::{Parse, ParseStream},
    punctuated::Punctuated,
    DeriveInput, Error as SynError, Generics, Ident, MetaList, Result as SynResult, Token,
};

static TOO_FEW_TYPES_ERR_MSG: &str = "at least two types required";

/// The input to the [`crate::Transitive`] derive macro.
pub struct TransitiveInput {
    ident: Ident,
    generics: Generics,
    paths: Vec<TransitionPath>,
}

impl TransitiveInput {
    const ATTR_NAME: &'static str = "transitive";
}

impl Parse for TransitiveInput {
    fn parse(input: ParseStream) -> SynResult<Self> {
        let DeriveInput {
            attrs,
            ident,
            generics,
            ..
        } = DeriveInput::parse(input)?;

        let fold_fn = |mut vec: Vec<TransitionPath>, res| {
            vec.extend(res?);
            Ok(vec)
        };

        let paths = attrs
            .into_iter()
            .filter(|a| a.path().is_ident(Self::ATTR_NAME))
            .map(|a| a.parse_args_with(Punctuated::<_, Token![,]>::parse_terminated))
            .try_fold::<_, _, SynResult<_>>(Vec::new(), fold_fn)?;

        let output = Self {
            ident,
            generics,
            paths,
        };

        Ok(output)
    }
}

impl ToTokens for TransitiveInput {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        for path in &self.paths {
            TokenizablePath::new(&self.ident, &self.generics, path).to_tokens(tokens);
        }
    }
}

/// Enum representing a path to take when transitioning from one type to another.
pub enum TransitionPath {
    From(TransitionFrom),
    Into(TransitionInto),
    TryFrom(TryTransitionFrom),
    TryInto(TryTransitionInto),
}

impl TransitionPath {
    const FROM: &'static str = "from";
    const INTO: &'static str = "into";
    const TRY_FROM: &'static str = "try_from";
    const TRY_INTO: &'static str = "try_into";
}

impl Parse for TransitionPath {
    fn parse(input: ParseStream) -> SynResult<Self> {
        let MetaList { path, tokens, .. } = MetaList::parse(input)?;
        let tokens = tokens.into();

        match path.require_ident()? {
            ident if ident == Self::FROM => syn::parse(tokens).map(TransitionPath::From),
            ident if ident == Self::INTO => syn::parse(tokens).map(TransitionPath::Into),
            ident if ident == Self::TRY_FROM => syn::parse(tokens).map(TransitionPath::TryFrom),
            ident if ident == Self::TRY_INTO => syn::parse(tokens).map(TransitionPath::TryInto),
            ident => Err(SynError::new(ident.span(), "unknown parameter")),
        }
    }
}

impl ToTokens for TokenizablePath<'_, &TransitionPath> {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        match &self.path {
            TransitionPath::From(from) => {
                TokenizablePath::new(self.ident, self.generics, from).to_tokens(tokens)
            }
            TransitionPath::Into(into) => {
                TokenizablePath::new(self.ident, self.generics, into).to_tokens(tokens)
            }
            TransitionPath::TryFrom(try_from) => {
                TokenizablePath::new(self.ident, self.generics, try_from).to_tokens(tokens)
            }
            TransitionPath::TryInto(try_into) => {
                TokenizablePath::new(self.ident, self.generics, try_into).to_tokens(tokens)
            }
        }
    }
}

/// Wrapper type that aids in the tokenization of [`TransitionPath`] and its variants.
pub struct TokenizablePath<'a, T> {
    pub ident: &'a Ident,
    pub generics: &'a Generics,
    pub path: T,
}

impl<'a, T> TokenizablePath<'a, T> {
    pub fn new(ident: &'a Ident, generics: &'a Generics, path: T) -> Self {
        Self {
            ident,
            generics,
            path,
        }
    }
}
