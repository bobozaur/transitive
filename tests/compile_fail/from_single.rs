//! Test that `#[transitive(from(T))]` should fail to compile

fn main() {
    use transitive::Transitive;
    
    struct A;

    struct B(A);

    impl From<A> for B {
        fn from(inner: A) -> Self {
            Self(inner)
        }
    }
    
    #[derive(Transitive)]
    #[transitive(from(A))] // impl From<A> for C
    struct C(B);

    impl From<B> for C {
        fn from(inner: B) -> Self {
            Self(inner)
        }
    }

    // Would cause an infinite loop
    let _: C = C::from(A);
}
