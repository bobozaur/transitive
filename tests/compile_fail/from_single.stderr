error: At least two types are required for a transitive conversion
  --> tests/compile_fail/from_single.rs:17:23
   |
17 |     #[transitive(from(A))] // impl From<A> for C
   |                       ^

error[E0277]: the trait bound `C: From<A>` is not satisfied
  --> tests/compile_fail/from_single.rs:27:16
   |
27 |     let _: C = C::from(A);
   |                ^ the trait `From<A>` is not implemented for `C`
   |
   = help: the trait `From<A>` is not implemented for `C`
           but trait `From<B>` is implemented for it
   = help: for that trait implementation, expected `B`, found `A`
